File: include/gui/IGUIObj.r
#ifndef _IGUIOBJ_R_
#define _IGUIOBJ_R_

/*
 * update:
 *  - If the returned value is not needed
 *    by convention function should return
 *    (int) 0
 */

#include "IClass.r"

#define CONT_IGUIOBJ \
	CONT_ICLASS \
	int (*update) ( void *obj ); \
	void (*draw) ( void *obj );

struct IGUIObj {
	CONT_IGUIOBJ
};

#endif

File: src/gui/IGUIObj.c
#include "gui/IGUIObj.r"
#include "gui/IGUIObj.h"

#include <assert.h>
#include "debug.h"

int update ( void *_self ) {
	const struct IGUIObj **self = _self;
	int ret = 0;

	assert ( *self );
	if ( (*self)->update ) {
		ret = (*self)->update ( _self );
	}

	return ret;
}

void draw ( void *_self ) {
	const struct IGUIObj **self = _self;

	assert ( *self );
	if ( (*self)->draw ) {
		(*self)->draw ( _self );
	}
}
